### PR TITLE
Feature/service abstraction

### PR DESCRIPTION
**Pull request by wim07101993**

### CODE REVIEW

File: Singer.API/Controllers/DataControllerBase.cs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+      /// The POST method for the controller to create a new entity in the database.
+      /// </summary>
+      /// <param name="dto">The new element to create in the database.</param>
+      /// <returns>The new created entity.</returns>
+      [HttpPost]
+      [ProducesResponseType(StatusCodes.Status201Created)]
+      [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+      public async Task<ActionResult<TDTO>> Create(TDTO dto)
+      {
+         var returnUser = await DatabaseService.CreateAsync(dto);
```
Comment (vyruz1986) [inline Singer.API/Controllers/DataControllerBase.cs]: Should be `returnItem` or similar? This is not specific for users

File: Singer.API/Controllers/DataControllerBase.cs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      /// <summary>
+      /// The GET method to get a single entity from the database.
+      /// </summary>
+      /// <param name="id">Id of the element to return.</param>
+      /// <returns>The entity in the database that has the given id.</returns>
+      [HttpGet("{id}")]
+      [ProducesResponseType(StatusCodes.Status200OK)]
+      [ProducesResponseType(StatusCodes.Status404NotFound)]
+      [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+      public virtual async Task<ActionResult<CareUserDTO>> GetOne(Guid id)
```
Comment (vyruz1986) [inline Singer.API/Controllers/DataControllerBase.cs]: Should be `TDTO`? Not user-specific

File: Singer.API/Controllers/DataControllerBase.cs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+      /// Updates a single entity in the database.
+      /// </summary>
+      /// <param name="id">The id of the entity to update.</param>
+      /// <param name="dto">The new value of the entity.</param>
+      /// <returns></returns>
+      [HttpPut("{id}")]
+      [ProducesResponseType(StatusCodes.Status200OK)]
+      [ProducesResponseType(StatusCodes.Status404NotFound)]
+      [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+      public async Task<ActionResult> UpdateUser(Guid id, TDTO dto)
```
Comment (vyruz1986) [inline Singer.API/Controllers/DataControllerBase.cs]: Should be `UpdateItem` or similar?

File: Singer.API/Controllers/DataControllerBase.cs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+      /// <summary>
+      /// Deletes a single entity from the database.
+      /// </summary>
+      /// <param name="id">The id of the entity to remove.</param>
+      /// <returns></returns>
+      [HttpDelete("{id}")]
+      [ProducesResponseType(StatusCodes.Status204NoContent)]
+      [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+      public async Task<ActionResult> DeleteUser(Guid id)
```
Comment (vyruz1986) [inline Singer.API/Controllers/DataControllerBase.cs]: should be `DeleteItem`, not user specific

**Review by vyruz1986 — CHANGES_REQUESTED** [review]
Please change some of the user-specific leftovers to generic names and types.
Fix the build
Also, 
Thanks

File: Singer.API/Controllers/DataControllerBase.cs (hunk 1 above)
Comment (wim07101993) [reply Singer.API/Controllers/DataControllerBase.cs]: fixin it

File: Singer.API/Controllers/DataControllerBase.cs (hunk 2 above)
Comment (wim07101993) [reply Singer.API/Controllers/DataControllerBase.cs]: fixin it

File: Singer.API/Controllers/DataControllerBase.cs (hunk 3 above)
Comment (wim07101993) [reply Singer.API/Controllers/DataControllerBase.cs]: fixed it

File: Singer.API/Controllers/DataControllerBase.cs (hunk 4 above)
Comment (wim07101993) [reply Singer.API/Controllers/DataControllerBase.cs]: fixed it

**Review by vyruz1986 — APPROVED** [review]

**Comment by vyruz1986** [comment]
Resolve conflicts & fix build. Code looks fine to me

File: Singer.API/Helpers/Extensions/IQueryableExtensions.cs [hunk 5]
```diff
@@ -95,5 +98,37 @@ public static IQueryable<T> TakePage<T>(this IOrderedQueryable<T> orderedQueryab
...
+               ? projection.OrderBy(orderByLambda)
+               : projection.OrderByDescending(orderByLambda);
+
+            items = await orderedQueryable.TakePage(pageIndex, pageSize).ToListAsync();
+         }
+
+         var result = new SearchResults<TProjection>(items, totalItemsCount, pageIndex);
+         return result;
+      }
    }
```
Comment (BerendWouters) [inline Singer.API/Helpers/Extensions/IQueryableExtensions.cs]: You could directly return this method.

File: Singer.API/Services/DatabaseService.cs [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+      #region CONSTRUCTOR
+
+      /// <summary>
+      /// Constructs a new instance of the <see cref="DatabaseService{TEntity, TDTO}"/> class.
+      /// </summary>
+      /// <param name="appDbContext">The context in which the database is approachable.</param>
+      /// <param name="mapper">The mapper to map the <see cref="TEntity"/>s to <see cref="TDTO"/>s and vice versa.</param>
+      protected DatabaseService(ApplicationDbContext appDbContext, IMapper mapper)
+      {
+         AppDbContext = appDbContext;
```
Comment (BerendWouters) [inline Singer.API/Services/DatabaseService.cs]: Since this is the only dbContext, it should suffice to name it `DbContext` (or even `Context`)

**Review by BerendWouters — APPROVED** [review]

**Pull request merged**